lib/cudnn/nnlib.jl

```diff
@@ -13,259 +13,225 @@ const CUDNNFloat = Union{Float16,Float32,Float64}
+    DenseConvDims{2,(1,K...),C_in,C_out,(1,S...),(0,0,P...),(1,D...),F}((1,cdims.I...))
 fix1d(pdims::PoolDims{1,K,S,P,D}) where {K,S,P,D,F} =
-  PoolDims{2,(K...,1),(S...,1),(P...,0,0),(D...,1)}((pdims.I..., 1), pdims.C_in)
-
-# We have to reshape the CuArray/PoolDims/DenseConvDims to 4D before feeding to CUDNN.
-reshape4D(x::AbstractVector) = reshape(x, 1, 1, length(x), 1)
-reshape4D(x::AbstractMatrix) = reshape(x, 1, 1, size(x)...)
-reshape4D(x::AbstractArray{T,3}) where T = reshape(x, size(x, 1), 1, size(x, 2), size(x, 3))
-reshape4D(x::AbstractArray{T}) where T = x
-
-workspacesize(x) = min(Mem.info()[1] ÷ 16, sizeof(x) * 2)
-
-function perfChoose(perfResults, returnedAlgoCount)::UInt32
-  if perfResults[1].status != 0
-    return 0
-  else
-    (best_algo,best_time,best_memory) = (perfResults[1].algo,perfResults[1].time,perfResults[1].memory)
-    for i = 2:returnedAlgoCount
-      if perfResults[i].status == 0 && perfResults[i].memory < best_memory && perfResults[i].time < best_time * 1.1
-        (best_algo,best_memory) = (perfResults[i].algo,perfResults[i].memory)
-      end
-    end
-    return best_algo
-  end
-end
-
+    PoolDims{2,(1,K...),(1,S...),(0,0,P...),(1,D...)}((1,pdims.I...), pdims.C_in)
 
 # Softmax
 
-# in-place for x or dy
+# @denizyuret: do not do inplace operations when (1) cpu version is not, (2) one can use softmax!
+# @denizyuret: recalculating y in ∇softmax! is a big waste, the nnlib API should be changed.
 softmax(x::DenseCuArray{T}; dims=1) where T<:CUDNNFloat =
-  softmax!(x, x, dims=dims)
+    softmax!(similar(x), x, dims=dims)
 
 ∇softmax(dy::DenseCuArray{T}, x::DenseCuArray{T}; dims=1) where T<:CUDNNFloat =
-  ∇softmax!(dy, dy, x, dims=dims)
+    ∇softmax!(similar(dy), dy, x, dims=dims)
 
 logsoftmax(x::DenseCuArray{T}; dims=1) where T<:CUDNNFloat =
-  logsoftmax!(x, x, dims=dims)
+    logsoftmax!(similar(x), x, dims=dims)
 
 ∇logsoftmax(dy::DenseCuArray{T}, x::DenseCuArray{T}; dims=1) where T<:CUDNNFloat =
-  ∇logsoftmax!(dy, dy, x, dims=dims)
+    ∇logsoftmax!(similar(dy), dy, x, dims=dims)
+
+function softmaxhelper(xsize, dims)
+    if length(xsize) == 1
+        xsize = (xsize..., 1)
+    end
+    if dims == (:)
+        return (CUDNN_SOFTMAX_MODE_INSTANCE, (1,1,prod(xsize),1))
+    elseif dims == 1
+        return (CUDNN_SOFTMAX_MODE_INSTANCE, (1,1,prod(xsize[1:end-1]),xsize[end]))
+    elseif dims == 2
+        return (CUDNN_SOFTMAX_MODE_CHANNEL, (1,prod(xsize[1:end-1]),xsize[end],1))
+    else
+        error("Don't know how to softmax with dims=$dims")
+    end
+end
 
 function softmax!(y::DenseCuArray{T}, x::DenseCuArray{T}; dims=1) where T<:CUDNNFloat
-  cudnnSoftmaxForward(reshape4D(x), reshape4D(y),
-                      algo=CUDNN_SOFTMAX_FAST, mode=cudnnSoftmaxMode_t(dims-1))
-  return y
+    mode, xsize = softmaxhelper(size(x), dims)
+    cudnnSoftmaxForward!(reshape(y,xsize), reshape(x,xsize); algo=CUDNN_SOFTMAX_FAST, mode)
```

Comment:
Does this always use fast mode? Would it be possible ti respect the default math mode here?

Reply:
It does right now, but I agree it should be consistent. I am not sure what the "default math mode" should be here: there is a CUDNN.math_mode, but that's about tensor optimizations.

Reply:
`CUDA.math_mode` is a more generic property, which `CUDNN.math_mode` copies for handle creation. But the former should be used for other purposes:
https://github.com/JuliaGPU/CUDA.jl/blob/d16a4b3194d29e1473dde5edb53a164045c4be33/lib/cublas/wrappers.jl#L757-L763

Reply:
I think we should have the current math mode as the default and have libraries request the fast paths where possible.